File: electron/duplicates.ts
import { createHash } from 'node:crypto'
import { createReadStream, promises as fs } from 'node:fs'
import path from 'node:path'
import type { DuplicateAnalysisResult, DuplicateFile, DuplicateGroup, DuplicateProgress } from './types.js'

const HASH_CONCURRENCY = 3

const fingerprint = (size: number, mtimeMs: number) => `${size}:${Math.trunc(mtimeMs)}`

async function hashFile(file: DuplicateFile, signal: AbortSignal, onBytes: (bytes: number) => void) {
  const hash = createHash('sha256')
  const stream = createReadStream(file.path, { highWaterMark: 1024 * 1024 })
  const abort = () => stream.destroy(new Error('Duplicate analysis cancelled'))
  signal.addEventListener('abort', abort, { once: true })
  try {
    for await (const chunk of stream) {
      if (signal.aborted) throw new Error('Duplicate analysis cancelled')
      hash.update(chunk as Buffer)
      onBytes((chunk as Buffer).length)
    }
    return hash.digest('hex')
  } finally { signal.removeEventListener('abort', abort) }
}

export async function analyzeDuplicates(rootPath: string, signal: AbortSignal, progress: (value: DuplicateProgress) => void): Promise<DuplicateAnalysisResult> {
  const files: DuplicateFile[] = []
  const physicalFiles = new Set<string>()
  const pending = [path.resolve(rootPath)]
  let discovered = 0
  while (pending.length) {
    if (signal.aborted) throw new Error('Duplicate analysis cancelled')
    const target = pending.pop()!
    let stat
    try { stat = await fs.lstat(target) } catch { continue }
    if (stat.isSymbolicLink()) continue
    if (stat.isDirectory()) {
      let entries: string[]
      try { entries = await fs.readdir(target) } catch { continue }
      for (const entry of entries) pending.push(path.join(target, entry))
      continue
    }
    if (!stat.isFile()) continue
    discovered++
    if (stat.size > 0) {
      const physicalId = `${stat.dev}:${stat.ino}`
      if (!physicalFiles.has(physicalId)) {
        physicalFiles.add(physicalId)
        files.push({
          name: path.basename(target), path: target, parentPath: path.dirname(target), size: stat.size,
          modifiedAt: stat.mtime.toISOString(),
          createdAt: stat.birthtimeMs > 0 ? stat.birthtime.toISOString() : undefined,
          fingerprint: fingerprint(stat.size, stat.mtimeMs),
        })
      }
    }
    if (discovered % 100 === 0) progress({ phase: 'discovering', currentPath: target, filesProcessed: discovered, bytesHashed: 0 })
  }

  const bySize = new Map<number, DuplicateFile[]>()
  progress({ phase: 'discovering', currentPath: rootPath, filesProcessed: discovered, bytesHashed: 0 })
  for (const file of files) bySize.set(file.size, [...(bySize.get(file.size) ?? []), file])
  const candidates = [...bySize.values()].filter((group) => group.length > 1).flat()
  const totalBytes = candidates.reduce((sum, file) => sum + file.size, 0)
  let next = 0, hashed = 0, bytesHashed = 0
  const matches = new Map<string, { hash: string; files: DuplicateFile[] }>()
  const worker = async () => {
    while (next < candidates.length) {
      const file = candidates[next++]
      let digest: string
      try { digest = await hashFile(file, signal, (bytes) => { bytesHashed += bytes }) } catch (error) {
        if (signal.aborted) throw error
        continue
      }
      hashed++
      const key = `${file.size}:${digest}`
      const match = matches.get(key) ?? { hash: digest, files: [] }
      match.files.push(file); matches.set(key, match)
      progress({ phase: 'hashing', currentPath: file.path, filesProcessed: hashed, totalFiles: candidates.length, bytesHashed, totalBytes })
    }
  }
  await Promise.all(Array.from({ length: Math.min(HASH_CONCURRENCY, candidates.length) }, worker))
  const groups: DuplicateGroup[] = [...matches.values()].filter((match) => match.files.length > 1).map((match) => ({
    id: `${match.files[0].size}:${match.hash}`,
    size: match.files[0].size,
    hash: match.hash,
    wastedSpace: match.files[0].size * (match.files.length - 1),
    files: match.files.sort((a, b) => a.path.localeCompare(b.path)),
  })).sort((a, b) => b.wastedSpace - a.wastedSpace)
  return {
    groups,
    totalWastedSpace: groups.reduce((sum, group) => sum + group.wastedSpace, 0),
    duplicateFileCount: groups.reduce((sum, group) => sum + group.files.length - 1, 0),
    scannedFileCount: discovered,
    hashedFileCount: hashed,
  }
}

export async function fileMatches(file: DuplicateFile) {
  try {
    const stat = await fs.lstat(file.path)
    return stat.isFile() && !stat.isSymbolicLink() && fingerprint(stat.size, stat.mtimeMs) === file.fingerprint
  } catch { return false }
}

File: src-tauri/src/duplicates.rs
use crate::types::*;
use chrono::{DateTime, Utc};
use serde::Serialize;
use sha2::{Digest, Sha256};
use std::{
    collections::{HashMap, HashSet},
    fs::{self, File},
    io::Read,
    path::{Path, PathBuf},
    sync::atomic::{AtomicBool, Ordering},
};
use tauri::{AppHandle, Emitter};

#[derive(Clone, Serialize)]
#[serde(rename_all = "camelCase")]
struct Progress {
    phase: String,
    current_path: String,
    files_processed: usize,
    #[serde(skip_serializing_if = "Option::is_none")]
    total_files: Option<usize>,
    bytes_hashed: u64,
    #[serde(skip_serializing_if = "Option::is_none")]
    total_bytes: Option<u64>,
}

#[cfg(unix)]
fn physical(_: &Path, meta: &fs::Metadata) -> Option<(u64, u64)> {
    use std::os::unix::fs::MetadataExt;
    Some((meta.dev(), meta.ino()))
}
#[cfg(windows)]
fn physical(path: &Path, _: &fs::Metadata) -> Option<same_file::Handle> {
    same_file::Handle::from_path(path).ok()
}
fn iso(time: std::time::SystemTime) -> String {
    DateTime::<Utc>::from(time).to_rfc3339()
}
fn fingerprint(meta: &fs::Metadata) -> String {
    let modified = meta
        .modified()
        .ok()
        .and_then(|v| v.duration_since(std::time::UNIX_EPOCH).ok())
        .map(|v| v.as_millis())
        .unwrap_or(0);
    format!("{}:{modified}", meta.len())
}
pub fn file_matches(file: &DuplicateFile) -> bool {
    fs::symlink_metadata(&file.path)
        .map(|m| m.is_file() && !m.file_type().is_symlink() && fingerprint(&m) == file.fingerprint)
        .unwrap_or(false)
}

pub fn analyze(
    root: String,
    app: AppHandle,
    cancel: &AtomicBool,
) -> Result<DuplicateAnalysisResult, String> {
    cancel.store(false, Ordering::Relaxed);
    let mut files = Vec::new();
    let mut seen = HashSet::new();
    let mut pending = vec![PathBuf::from(&root)];
    let mut discovered: usize = 0;
    while let Some(target) = pending.pop() {
        if cancel.load(Ordering::Relaxed) {
            return Err("Duplicate analysis cancelled".into());
        }
        let meta = match fs::symlink_metadata(&target) {
            Ok(v) => v,
            Err(_) => continue,
        };
        if meta.file_type().is_symlink() {
            continue;
        }
        if meta.is_dir() {
            if let Ok(entries) = fs::read_dir(&target) {
                for entry in entries.flatten() {
                    pending.push(entry.path())
                }
            }
            continue;
        }
        if !meta.is_file() {
            continue;
        }
        discovered += 1;
        if meta.len() > 0 && physical(&target, &meta).is_none_or(|identity| seen.insert(identity)) {
            files.push(DuplicateFile {
                name: target
                    .file_name()
                    .unwrap_or_default()
                    .to_string_lossy()
                    .into_owned(),
                path: target.to_string_lossy().into_owned(),
                parent_path: target
                    .parent()
                    .unwrap_or(Path::new(""))
                    .to_string_lossy()
                    .into_owned(),
                size: meta.len(),
                modified_at: meta.modified().map(iso).unwrap_or_default(),
                created_at: meta.created().ok().map(iso),
                fingerprint: fingerprint(&meta),
            })
        }
        if discovered.is_multiple_of(100) {
            let _ = app.emit(
                "duplicate-progress",
                Progress {
                    phase: "discovering".into(),
                    current_path: target.to_string_lossy().into_owned(),
                    files_processed: discovered,
                    total_files: None,
                    bytes_hashed: 0,
                    total_bytes: None,
                },
            );
        }
    }
    let mut by_size: HashMap<u64, Vec<DuplicateFile>> = HashMap::new();
    for file in files {
        by_size.entry(file.size).or_default().push(file)
    }
    let candidates: Vec<_> = by_size
        .into_values()
        .filter(|v| v.len() > 1)
        .flatten()
        .collect();
    let total_bytes = candidates.iter().map(|f| f.size).sum();
    let mut matches: HashMap<(u64, String), Vec<DuplicateFile>> = HashMap::new();
    let mut bytes_hashed = 0;
    let mut hashed = 0;
    for file in candidates.iter() {
        if cancel.load(Ordering::Relaxed) {
            return Err("Duplicate analysis cancelled".into());
        }
        let mut input = match File::open(&file.path) {
            Ok(v) => v,
            Err(_) => continue,
        };
        let mut digest = Sha256::new();
        let mut buffer = vec![0u8; 1024 * 1024];
        loop {
            let count = input.read(&mut buffer).map_err(|e| e.to_string())?;
            if count == 0 {
                break;
            }
            digest.update(&buffer[..count]);
            bytes_hashed += count as u64;
            if cancel.load(Ordering::Relaxed) {
                return Err("Duplicate analysis cancelled".into());
            }
        }
        hashed += 1;
        let hash = hex::encode(digest.finalize());
        matches
            .entry((file.size, hash))
            .or_default()
            .push(file.clone());
        let _ = app.emit(
            "duplicate-progress",
            Progress {
                phase: "hashing".into(),
                current_path: file.path.clone(),
                files_processed: hashed,
                total_files: Some(candidates.len()),
                bytes_hashed,
                total_bytes: Some(total_bytes),
            },
        );
    }
    let mut groups: Vec<_> = matches
        .into_iter()
        .filter(|(_, v)| v.len() > 1)
        .map(|((size, hash), mut files)| {
            files.sort_by(|a, b| a.path.cmp(&b.path));
            DuplicateGroup {
                id: format!("{size}:{hash}"),
                size,
                hash,
                wasted_space: size * (files.len() as u64 - 1),
                files,
            }
        })
        .collect();
    groups.sort_by_key(|group| std::cmp::Reverse(group.wasted_space));
    Ok(DuplicateAnalysisResult {
        total_wasted_space: groups.iter().map(|g| g.wasted_space).sum(),
        duplicate_file_count: groups.iter().map(|g| g.files.len() - 1).sum(),
        groups,
        scanned_file_count: discovered,
        hashed_file_count: hashed,
    })
}
